Payments: Improve manage purchase UI so that users are encourage to renew, not just update their payment method

DavidRothstein opened the pull request:
Note: This pull request depends on three others (all of which have since been merged):
- https://github.com/Automattic/wp-calypso/pull/26429
- https://github.com/Automattic/wp-calypso/pull/26448
- https://github.com/Automattic/wp-calypso/pull/26480

## The problem

There have been reports (p2MSmN-5aw-p2) that some users are updating their payment method for a subscription under the assumption that that will also cause the subscription to immediately renew.  (One possible reason they might do this is if they don't realize that clicking "Renew Now" also will give you an opportunity to use a new payment method, so they start by adding a payment method instead and then get confused afterwards.  A second possible reason is that in some cases, they don't notice the "Renew Now" button at all.)

These users will eventually be auto-renewed, but since it is confusing them that the subscription wasn't renewed right away, and since we'd like to encourage more people to go ahead and actually try the renewal right after adding a new credit card (to be able to make sure that the card can be charged with no problem) we'd like to fix this.

## Solution

This issue proposes two fixes to the subscription Manage Purchase page for the case of subscriptions that are up for renewal:
- Add a "Renew Now" link right above the "Edit Payment Method" nav item.
- If the user goes through the process of adding/editing a credit card, make the message they see after the card was added explain that the renewal hasn't happened yet, and give them a link to renew now there too.

## How to test

Add a subscription with an expiration date such it will expire soon (in the next 90 days) and then navigate to the Manage Purchase page and look for/test the new options shown in the screenshots below.

Or, add a subscription with an expiration date farther in the future, and verify that the patch doesn't change anything on the Manage Purchase page for it.

## Screenshots

### If the subscription is up for renewal soon

There's a new "Renew Now" link as shown by the arrow in the screenshot below.  (The other "Renew Now" button in this screenshot was already there before this pull request, and is more subtle in this example - although if it's expiring soon and not going to auto-renew it is in an error-style message, which is much louder.  This patch doesn't touch the existing link in either scenario, just adds the new one in the navigation below):

![renewing_soon_before_credit_card_change](https://user-images.githubusercontent.com/235183/43426670-849b59ac-9424-11e8-87c6-a8f07badd310.png)

After you click "Edit Payment Method" and successfully add a new credit card, the message you see on the screen after being redirected back to this page looks like the following (with a functional "Renew Now" link inside the message):

![renewing_soon_after_credit_card_change](https://user-images.githubusercontent.com/235183/43426677-8b89ad72-9424-11e8-87cd-3628f85f8c6d.png)

### If the subscription isn't up for renewal soon

There are no changes from before this pull request.  I'm showing screenshots here for comparison with the above.

The "Manage Purchase" page for the subscription looks like this:

![not_renewing_soon_before_credit_card_change](https://user-images.githubusercontent.com/235183/43426657-7ac9e66e-9424-11e8-90c0-84b38b0d791c.png)

With the following message after you change your payment method:

![not_renewing_soon_after_credit_card_change](https://user-images.githubusercontent.com/235183/43426664-7eb4494a-9424-11e8-9b37-31363dbf0f30.png)

DavidRothstein commented:
Note: This currently includes https://github.com/Automattic/wp-calypso/pull/26429 as part of it, since the changes to included files in this PR trigger that somewhat-unrelated bug (without including that, `npm start` fails to build).

However, it looks like there are some other possibly similar problems that will need to be fixed here too in order to get all the tests to pass.

DavidRothstein commented:
>However, it looks like there are some other possibly similar problems that will need to be fixed here too in order to get all the tests to pass.

This should be fixed now, and I expect the tests to pass.  I'll explain more in an inline comment.

DavidRothstein commented on client/lib/cart-values/test/index.js:
```diff
@@ -1,4 +1,7 @@
+/**
+ * @format
+ * @jest-environment jsdom
+ */
```
This change is needed because the shuffling around of include files in this PR caused `client/lib/analytics/index.js` to be loaded during this test, and that file assumes a browser environment when it's loaded.

We could fix it by fixing that library file instead, similar to https://github.com/Automattic/wp-calypso/pull/26429, but it looks like it would involve a larger and more complicated set of changes, so I felt it was simpler to just make the tests pass this way.  This PR can't be responsible for fixing every problem in Calypso :)

DavidRothstein commented on client/state/sites/domains/test/actions.js:
```diff
@@ -1,4 +1,7 @@
+/**
+ * @format
+ * @jest-environment jsdom
+ */
```
This change is needed for the same reason as the one in `client/lib/cart-values/test/index.js`.

Now that it turns out there are two cases of this, it makes me wonder a little more if actually fixing `client/lib/analytics/index.js` not to assume a browser environment when it's loaded is a better way to go... but still feels like it would be taking me pretty far outside the original goal of this PR :)

DavidRothstein commented:
Looks like I misread the CircleCI test results, and there are actually a lot of test failures here.  So I'm going to have to go ahead and fix `client/lib/analytics/index.js` for real, rather than fixing the tests one by one.

DavidRothstein commented:
This now includes https://github.com/Automattic/wp-calypso/pull/26448, which fixes `client/lib/analytics/index.js` to be loadable outside a browser environment.

However, it turns out that doing that is necessary, but not sufficient, to fix all the failing tests.  There are still failures (in some cases) or warnings (in other cases) due to "UnhandledPromiseRejectionWarning: Error: No available storage method found."

It looks like this is a common test error and the way it's been dealt with in the past (for example, in https://github.com/Automattic/wp-calypso/pull/25223) is to add this to the affected test file:

```
// Gets rid of warnings such as 'UnhandledPromiseRejectionWarning: Error: No available storage method found.'
jest.mock( 'lib/user', () => () => {} );
```

So I've done that to the remaining tests that had problems here, and hopefully that will fix all of the failures.

DavidRothstein commented:
The tests are (finally) now passing, all the `UnhandledPromiseRejectionWarning: Error: No available storage method found` messages in the tests are now gone, `npm start` works, and this is finally ready for actual review of the intended changes...

I'm adding @sirbrillig as a direct reviewer here, since he is the most likely to actually review it.  Thanks :)

sirbrillig commented on client/components/card/README.md:
```diff
@@ -32,7 +32,8 @@ Name | Type | Default | Description
 `className` | `string` | null | Adds CSS classes.
 `href` | `string` | null | URL of the card. Adds a right chevron icon.
-`tagName` | `string` | null | Allows you to control the tag name of the card wrapper (only if `href` is not specified).
+`fakeLink` | `bool` | false | Used for cards without a `href` that need to display as a link. The card will render as an accessible button with the same visual appearance as a link (including the right chevron icon).
```
🤔 this does seem like the simplest way to get this behavior. What if we called it `clickable` rather than `fakeLink`? To me that seems more clear; then `href` is more like a shortcut to adding `clickable` and a URL. It might be worth noting in the docs that if you specify `clickable` you should also have `href` or `onClick` set?

DavidRothstein replied:
Not sure about `clickable` because there can be clickable cards (e.g. regular button cards) that don't want this extra styling.

How about something like `displaysAsLink` instead, but also require the user to specify the tag?  So for example, `<Card tagName="button" displaysAsLink onClick=...` would make a button that looks like a link.

> It might be worth noting in the docs that if you specify clickable you should also have href or onClick set?

Agreed, since it's primarily for buttons, I'll definitely mention the `onClick`.

sirbrillig commented:
Looking good. I am not sure about the `fakeLink` flag; left an inline comment above.

DavidRothstein commented:
I switched from `fakeLink` to `displaysAsLink` (as discussed in the inline comments), and also updated this pull request to base it off the latest code in master, since the two pull requests it depended on are now in.

Does it look good now?  Thanks for the review!

DavidRothstein commented:
Makes sense (on both counts).  I updated it to `displayAsLink` now, and I'll copy the card-related changes to a new PR.  That way there's something tracking it separately, and if the renewal-related changes here need to be reverted for some reason, the new `Card` property is still available.
